From a47fe9c151fdd6bea2926e853f61521d5471bf98 Mon Sep 17 00:00:00 2001
From: hmgokhale <hmgokhale@maxlinear.com>
Date: Tue, 8 Oct 2024 18:20:12 +0530
Subject: [PATCH] WLANRTSYS-85449 Ignore assoc_cb if STA is already associated

Issue: Call Trace seen while changing bss color , channel and bw using hostapd_cli cmd
RCA:
The duplicate assoc_resp_cb is not actually a duplicate event but after first association there is
a timeout and the there is a disconnect event as the assoc_resp_cb is not received.
Second time when there is assoc and eapol is completed we are getting the old assoc_resp_cb which is
causing the fw fatal and assert due to wrong removal of STA.

Fix: Ignore assoc_resp_cb if STA is already associated

Fixup to WLANRTSYS-62845 MLO changes in iwlwav-hostap-ng: AP MLD Auth Assoc and eapol changes

Signed-off-by: hmgokhale <hmgokhale@maxlinear.com>
---
 src/ap/ieee802_11.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index dc27bcfa4..5b255d7a0 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7117,6 +7117,15 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 	}
 #endif /* CONFIG_IEEE80211BE */
 
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+	if (mxl_is_mlo_enabled(hapd->conf) && mxl_mlo_is_ml_sta(sta)) {
+		if (!reassoc && (sta->flags & WLAN_STA_ASSOC)) {
+			wpa_printf(MSG_DEBUG, "Ignore the assoc_response_cb as it is already processed");
+			return;
+		}
+	}
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
+
 	if (len < IEEE80211_HDRLEN + (reassoc ? sizeof(mgmt->u.reassoc_resp) :
 				      sizeof(mgmt->u.assoc_resp))) {
 		wpa_printf(MSG_INFO,
-- 
2.43.0

